sycl/test-e2e/lit.cfg.py

```diff
@@ -178,6 +178,63 @@
 else:
     config.substitutions.append( ('%cuda_options', '') )
 
+# Check for device features
+device_features_file='device_features.cpp'
+with open(device_features_file, 'w') as fp:
+    fp.write("""
+#include <sycl/sycl.hpp>
+
+int main() {
+  using namespace sycl;
+
+  auto all_devices = device::get_devices();
+  if (all_devices.size() == 0)
+    return 0;
+
+  auto all_of_devices = [&](auto Pred) {
+    return std::all_of(all_devices.begin(), all_devices.end(), Pred);
+  };
+
+  auto all_devices_have_aspect = [&](aspect an_aspect) {
+    return all_of_devices([&an_aspect](const device &dev) {
+      try {
+        return dev.has(an_aspect);
+      } catch (...) {
+        return false;
+      }
+    });
+  };
+#define PRINT_IF_ALL_HAVE(ASPECT)                                              \
+  if (all_devices_have_aspect(aspect::ASPECT))                                 \
+    std::cout << "aspect_" #ASPECT << std::endl;
+
+#define __SYCL_ASPECT(ASPECT, ID) PRINT_IF_ALL_HAVE(ASPECT)
+#include <sycl/info/aspects.def>
+
+  auto is_pvc = [&](const device &dev) {
+    if (dev.has(aspect::ext_intel_device_id)) {
+      auto device_id = dev.get_info<ext::intel::info::device::device_id>();
+      return (device_id & 0xff0) == 0xbd0;
+    }
+    return false;
+  };
+  if (all_of_devices(is_pvc))
+    std::cout << "gpu-intel-pvc" << std::endl;
+  return 0;
+}
+""")
+
+cmd = (config.dpcpp_compiler+' -fsycl  ' + device_features_file + ' ' +
+       ('/Fe' if cl_options else '-o ') + ' device_features.exe')
```

Comment:
Are all these commands executed for each test run?

Reply:
I hope not, because the code above/below this new snippet is doing the same thign, basically.

Reply:
Just from this code snippet, I don't see any optimizations caching the result of these commands. I assume we write the file, compile it and run the executable each time test is executed.

https://llvm.org/docs/CommandGuide/lit.html#test-suites

Reply:
From the link you've mentioned:

> lit identifies test suites as directories containing lit.cfg or lit.site.cfg files

> Once a test suite is discovered, its config file is loaded. Config files themselves are Python modules which will be executed.

In other words, it's done once per llvm-lit invocation. With extra print from this code I'm getting something like

```
$ llvm-lit -sv --param sycl_be=level_zero --param target_devices=gpu .
# The "device_features" check is executed here
...
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:293: note: Backend: ext_oneapi_level_zero
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:374: warning: CPU device not used
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:389: note: Test GPU device
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:432: warning: Accelerator device not used
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:467: note: Found llvm-spirv
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:467: note: Found llvm-link
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:483: note: Found pre-installed AOT device compiler ocloc
llvm-lit: <...>/llvm/sycl/test-e2e/lit.cfg.py:483: note: Found pre-installed AOT device compiler opencl-aot
                           -- Testing: 3 tests, 3 workers --

FAIL: SYCL :: TestSuiteMove/fail.cpp (1 of 3)
******************** TEST 'SYCL :: TestSuiteMove/fail.cpp' FAILED ********************
Script:
--
: 'RUN: at line 1';   false
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 1"
note: command had no output on stdout or stderr
$ "false"
note: command had no output on stdout or stderr
error: command failed with exit status: 1

--

********************
                           -- Testing: 3 tests, 3 workers --

********************
Failed Tests (1):
  SYCL :: TestSuiteMove/fail.cpp


Testing Time: 4.79s
  Passed: 2
  Failed: 1

```

Reply:
> In other words, it's done once per llvm-lit invocation.

That's my understanding as well and it doesn't seem to be optimal.